src/mem/pool.h

```diff
@@ -32,14 +36,88 @@ namespace snmalloc
     constexpr PoolState() = default;
   };
 
-  template<typename T>
+  /**
+   * Helper class used to instantiate a global PoolState.
+   *
+   * SingletonPoolState::pool is the default provider for the PoolState within
+   * the Pool class.
+   */
+  template<typename T, typename SharedStateHandle>
+  class SingletonPoolState
```

Comment:
Is it possible to use
https://github.com/microsoft/snmalloc/blob/6e638742e3c66549174d4c264bd05c9435938ac1/src/ds/helpers.h#L15-L46

Reply:
Initially I tried to use Singleton, but it assumes that the Object is copy assignable (obj = init()), which PoolState cannot be (ABA is not copy contructible and it is good that it is not). As I looked through the code, changing Singleton to use obejct constructors instead of init did not look viable as there are many use case with non-class types.

Reply:
Cool.  Thanks just wanted to check it had been considered.

Reply:
This seems like something that it would be good to fix in Singleton, since the ability to copy a singleton is not likely to be desirable.  I believe the copy here isn't actually a copy, it's expected to invoke guaranteed copy elision.  Can this be worked around by providing a private copy constructor with an `assert(0)` in it and making Singleton a friend?  If the copy constructor is actually executed, then we have a bug, but if guaranteed copy elision actually works then the object is created in place and the copy constructor is never called.